File: z_lc/207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        n = numCourses
        G = [set() for _ in range(n)]
        num_prereqs = [0] * n
        for course,prereq in prerequisites:
            G[prereq].add(course)
            num_prereqs[course] += 1
            
        can_take = [i for i in range(n) if num_prereqs[i] == 0]
        i = 0
        while i < len(can_take):
            taken = can_take[i]
            for postreq in G[taken]:
                num_prereqs[postreq] -= 1
                if num_prereqs[postreq] == 0:
                    can_take.append(postreq)
            i += 1
        return len(can_take) == n

File: z_lc/207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        int n = numCourses;
        vector<vector<int>> G(n);
        vector<int> num_prereqs(n, 0);
        for(auto& e : prerequisites) {
            int &course = e[0], &prereq = e[1];
            G[prereq].push_back(course);
            ++num_prereqs[course];
        }
        
        vector<int> can_take;
        for(int i=0; i<n; ++i) {
            if(num_prereqs[i] == 0) can_take.push_back(i);
        }
        
        for(int i=0; i<can_take.size(); ++i) {
            int taken = can_take[i];
            for(int& postreq : G[taken]) {
                if(--num_prereqs[postreq] == 0) can_take.push_back(postreq);
            }
        }
        return can_take.size() == n;
    }
};